exp/api/v1alpha3/types.go

```diff
@@ -83,3 +83,131 @@ var (
+type NetworkInterface struct {
+	DeviceIndex int64    `json:"deviceIndex,omitempty"`
+	Groups      []string `json:"groups,omitempty"`
+}
+
+// AwsLaunchTemplate defines the desired state of AWSLaunchTemplate
+type AWSLaunchTemplate struct {
+	// all the things needed for a launch template
+	ID   string `json:"id,omitempty"`
+	Name string `json:"name,omitempty"`
+
+	IamInstanceProfile string             `json:"iamInstanceProfile,omitempty"`
+	NetworkInterfaces  []NetworkInterface `json:"networkInterfaces,omitempty"`
+
+	// todo: use a helper
+	AMI infrav1.AWSResourceReference `json:"ami,omitempty"`
+
+	// ImageLookupFormat is the AMI naming format to look up the image for this
+	// machine It will be ignored if an explicit AMI is set. Supports
+	// substitutions for {{.BaseOS}} and {{.K8sVersion}} with the base OS and
+	// kubernetes version, respectively. The BaseOS will be the value in
+	// ImageLookupBaseOS or ubuntu (the default), and the kubernetes version as
+	// defined by the packages produced by kubernetes/release without v as a
+	// prefix: 1.13.0, 1.12.5-mybuild.1, or 1.17.3. For example, the default
+	// image format of capa-ami-{{.BaseOS}}-?{{.K8sVersion}}-* will end up
+	// searching for AMIs that match the pattern capa-ami-ubuntu-?1.18.0-* for a
+	// Machine that is targeting kubernetes v1.18.0 and the ubuntu base OS. See
+	// also: https://golang.org/pkg/text/template/
+	// +optional
+	ImageLookupFormat string `json:"imageLookupFormat,omitempty"`
+
+	// ImageLookupOrg is the AWS Organization ID to use for image lookup if AMI is not set.
+	ImageLookupOrg string `json:"imageLookupOrg,omitempty"`
+
+	// ImageLookupBaseOS is the name of the base operating system to use for
+	// image lookup the AMI is not set.
+	ImageLookupBaseOS string `json:"imageLookupBaseOS,omitempty"`
+
+	// InstanceType is the type of instance to create. Example: m4.xlarge
+	InstanceType string `json:"instanceType,omitempty"`
+
+	// RootVolume encapsulates the configuration options for the root volume
+	// +optional
+	RootVolume *infrav1.Volume `json:"rootVolume,omitempty"`
+
+	// SSHKeyName is the name of the ssh key to attach to the instance. Valid values are empty string (do not use SSH keys), a valid SSH key name, or omitted (use the default SSH key name)
+	// +optional
+	SSHKeyName *string `json:"sshKeyName,omitempty"`
+
+	VersionNumber *int64 `json:"versionNumber,omitempty"`
+
+	// AdditionalSecurityGroups is an array of references to security groups that should be applied to the
+	// instances. These security groups would be set in addition to any security groups defined
+	// at the cluster level or in the actuator.
+	// +optional
+	AdditionalSecurityGroups []infrav1.AWSResourceReference `json:"additionalSecurityGroups,omitempty"`
+}
+
+// Overrides from describe-auto-scaling-groups
+type Overrides struct {
+	InstanceType string `json:"instanceType"`
+}
+
+// InstancesDistribution from describe-auto-scaling-groups
+type InstancesDistribution struct {
```

Comment:
oh! this makes sense

Reply:
I'm not sure I see a benefit to using a custom type. `SpotAllocationStrategy` is just a string that goes into making a mixedInstancesPolicy [here](https://github.com/newrelic-forks/cluster-api-provider-aws/blob/machinepools/pkg/cloud/services/autoscaling/autoscalinggroup.go#L269)

Reply:
Normally the benefit comes when you need to do some conditional processing based on the value, so if you have a type then you deal in the type instances instead of having to remember or look up the possible string values. If its just a 1:1 into mixedInstancesPolicy with conditional then that benefit isn't needed.

Reply:
Yup. It's super helpful for downstream consumers of the API if they're importing the types.

Reply:
got it. this should be all set